### PR TITLE
ReHMC support for spectrahedra and general convex bodies with refactored Spectrahedra classes

### PR DESCRIPTION
**Pull request by papachristoumarios**
Combines the following 2 PRs on a new PR with a renamed branch: 

* https://github.com/papachristoumarios/volume_approximation/pull/2
* https://github.com/GeomScale/volesti/pull/159

### CODE REVIEW

File: include/convex_bodies/spectrahedra/spectrahedron.h [hunk 1]
```diff
@@ -104,62 +141,229 @@ class Spectrahedron {
...
 
         // create matrices A, B, C
-        createMatricesForPositiveIntersection(a, b, c, precomputedValues);
+        createMatricesForPositiveQuadIntersection(a, b, c);
 
         // get the minimum positive eigenvalue of At^2 + Bt + C
-        EigenvaluesProblems<NT, MT, VT> quadraticEigenvaluesProblem;
-        NT distance = quadraticEigenvaluesProblem.minPosQuadraticEigenvalue(precomputedValues.A, precomputedValues.B,
+        NT distance = EigenvaluesProblem.minPosQuadraticEigenvalue(precomputedValues.A, precomputedValues.B,
                                                                             precomputedValues.C, precomputedValues.X,
```
Comment (vissarion) [inline include/convex_bodies/spectrahedra/spectrahedron.h]: indentation seems wrong after this

File: include/convex_bodies/spectrahedra/spectrahedron.h [hunk 2]
```diff
@@ -14,49 +14,69 @@
...
 
 
+/// Among successive calls of this class methods, we may need to pass data
+/// from one call to the next, to avoid repeating computations, or to efficiently update values
+/// Warning: this struct assists in many methods; perhaps for different methods use different instances
+template <typename NT, typename MT, typename VT>
+struct PrecomputationOfValues {
+
+    /// These flags indicate whether the corresponding matrices are computed
+    /// if yes, we can use them and not compute them fro scratch
```
Comment (vissarion) [inline include/convex_bodies/spectrahedra/spectrahedron.h:24]: OK I understand that the use of flags is not the intent of this PR but why not including a TODO in the comments to avoid the use of flags in the future?

File: include/diagnostics/print_diagnostics.hpp [hunk 3]
```diff
@@ -28,6 +28,8 @@ void print_diagnostics(MT const& samples, unsigned int &min_ess, StreamType &str
     VT ess = effective_sample_size<NT, VT, MT>(samples, min_ess);
     VT intv_psrf = interval_psrf<VT, NT, MT>(samples);
 
+    //std::cout << "interval_psrf =  " << intv_psrf.maxCoeff() << "us" << std::endl;
```
Comment (vissarion) [inline include/diagnostics/print_diagnostics.hpp]: is this needed?

File: include/matrix_operations/EigenvaluesProblems.h [hunk 4]
```diff
@@ -14,18 +14,18 @@
...
 //#define EIGEN_EIGENVALUES_SOLVER
 /// Spectra standard eigenvalue problem
-//#define SPECTRA_EIGENVALUES_SOLVER
+#define SPECTRA_EIGENVALUES_SOLVER
 /// ARPACK++ standard eigenvalues solver
-#define ARPACK_EIGENVALUES_SOLVER
+//#define ARPACK_EIGENVALUES_SOLVER
 
-#include <../../external/arpack++/include/arssym.h>
+//#include <../../external/arpack++/include/arssym.h>
```
Comment (vissarion) [inline include/matrix_operations/EigenvaluesProblems.h]: leftover?

File: include/matrix_operations/EigenvaluesProblems.h [hunk 5]
```diff
@@ -14,18 +14,18 @@
 /// Eigen solver for generalized eigenvalue problem
 //#define EIGEN_EIGENVALUES_SOLVER
 /// Spectra standard eigenvalue problem
-//#define SPECTRA_EIGENVALUES_SOLVER
+#define SPECTRA_EIGENVALUES_SOLVER
 /// ARPACK++ standard eigenvalues solver
-#define ARPACK_EIGENVALUES_SOLVER
+//#define ARPACK_EIGENVALUES_SOLVER
```
Comment (vissarion) [inline include/matrix_operations/EigenvaluesProblems.h:19]: leftover?

File: include/matrix_operations/EigenvaluesProblems.h [hunk 6]
```diff
@@ -14,18 +14,18 @@
...
-#include <../../external/arpack++/include/arssym.h>
+//#include <../../external/arpack++/include/arssym.h>
 #include <../../external/Spectra/include/Spectra/SymEigsSolver.h>
 #include "DenseProductMatrix.h"
 #include "EigenDenseMatrix.h"
 
 #include "../../external/Spectra/include/Spectra/SymGEigsSolver.h"
 #include "../../external/Spectra/include/Spectra/GenEigsSolver.h"
-#include "../../external/arpack++/include/arsnsym.h"
+//#include "../../external/arpack++/include/arsnsym.h"
```
Comment (vissarion) [inline include/matrix_operations/EigenvaluesProblems.h]: leftover?

File: include/matrix_operations/EigenvaluesProblems.h [hunk 7]
```diff
@@ -287,6 +287,40 @@ class EigenvaluesProblems<NT, Eigen::Matrix<NT,Eigen::Dynamic,Eigen::Dynamic>, E
...
 
+    /// Find the minimum positive and maximum negative eigenvalues of the generalized eigenvalue
+    /// problem A + lB, where A, B symmetric and A negative definite.
+    /// \param[in] A Input matrix
+    /// \param[in] B Input matrix
+    /// \return The pair (minimum positive, maximum negative) of eigenvalues
+    NT minPosLinearEigenvalue(MT const & A, MT const & B, VT &eigvec) const
+    {
+        int matrixDim = A.rows();
+        double lambdaMinPositive;
```
Comment (vissarion) [inline include/matrix_operations/EigenvaluesProblems.h:296]: you can avoid declaration here (is it always `double` ?) and use `auto` in initialization below.

File: include/ode_solvers/generalized_leapfrog.hpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  typedef typename ConvexBody::VT VT;
+
+  unsigned int dim;
+  std::vector<NT> lambda_prev;
+
+
+  NT eta;
+  NT eta0;
+  NT t;
+  NT dl = 0.995;
```
Comment (vissarion) [inline include/ode_solvers/generalized_leapfrog.hpp:36]: `const`

File: include/random_walks/boltzmann_hmc_walk.hpp [hunk 9]
```diff
@@ -97,76 +88,40 @@ struct BoltzmannHMCWalk {
...
-
         /// Samples random points from the spectrahedron from the Boltzmann distribution
         /// \param[in] spectrahedron A spectrahedron
         /// \param[in] interiorPoint A point in the interior of the spectrahedron
         /// \param[in] pointsNum The number of points to sample
         /// \param[out] points The list of the sampled points
         /// \tparam Point class Point with NT and VT as declared above in this class
         template <typename Point>
-        void apply(SPECTRAHEDRON &spectrahedron, Point const & interiorPoint, const unsigned int pointsNum,
+        void apply(ConvexBody &spectrahedron, Point const & interiorPoint, const unsigned int pointsNum,
```
Comment (vissarion) [inline include/random_walks/boltzmann_hmc_walk.hpp]: the parameter name should be renamed as well, right? From `spectahedron` to `convexbody`

File: include/random_walks/boltzmann_hmc_walk.hpp [hunk 10]
```diff
@@ -97,76 +88,40 @@ struct BoltzmannHMCWalk {
...
 
         /// A single step of the HMC random walk: choose a direction and walk on the trajectory for a random distance.
         /// If it hits the boundary, the trajectory is reflected. If #reflections < reflectionsBound * dimension, it returns the same point
         /// \param[in] spectrahedron A spectrahedron
         /// \param[in, out] p An interior point, and the next point in the random walk
-        /// \param[in, out] precomputedValues Data for the methods of the class Spectrahedron
         /// \tparam Point
         template <typename Point>
-        void getNextPoint(SPECTRAHEDRON &spectrahedron, VT &p, PrecomputedValues &precomputedValues) {
+        void getNextPoint(ConvexBody &spectrahedron, VT &p) {
```
Comment (vissarion) [inline include/random_walks/boltzmann_hmc_walk.hpp]: here as above

File: test/LpSolve.cmake [hunk 11]
```diff
@@ -0,0 +1,41 @@
+function(GetLpSolve)
```
Comment (vissarion) [inline test/LpSolve.cmake]: is this lpsolve related change needed in the tests?

File: test/logconcave_sampling_test.cpp [hunk 12]
```diff
@@ -757,6 +880,81 @@ void call_test_benchmark_polytopes_grid_search() {
...
 
+
+template <typename NT>
+void call_test_benchmark_spectrahedra_grid_search() {
+
+    typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+    typedef Eigen::Matrix <NT, Eigen::Dynamic, Eigen::Dynamic> MT;
+    typedef Cartesian<NT> Kernel;
+    typedef typename Kernel::Point    Point;
+    typedef Spectrahedron <Point> SPECTRAHEDRON;
```
Comment (vissarion) [inline test/logconcave_sampling_test.cpp]: to inline with the current writing `typedef Spectrahedron<Point> Spectrahedron`

File: test/logconcave_sampling_test.cpp [hunk 13]
```diff
@@ -757,6 +880,81 @@ void call_test_benchmark_polytopes_grid_search() {
 
 }
 
+
+template <typename NT>
+void call_test_benchmark_spectrahedra_grid_search() {
+
+    typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+    typedef Eigen::Matrix <NT, Eigen::Dynamic, Eigen::Dynamic> MT;
```
Comment (vissarion) [inline test/logconcave_sampling_test.cpp]: no space needed after `Matrix`

**Review by vissarion — COMMENTED** [review]
Thanks for this PR! I have some comments. 

Regarding coverage I think you should ask more tests to cover your code, then codecov will not complain. 

Also, do you agree to merge https://github.com/GeomScale/volesti/pull/190 before this PR? Then your examples here could be simplified.

**Review by TolisChal — APPROVED** [review]

**Review by vissarion — COMMENTED** [review]
Thanks, I am OK with merging. Coverage can be improved, but in a future PR that will handle spectahedra better.
